### PR TITLE
8200224: Multiple press event when JFXPanel gains focus

### PR DESCRIPTION
**Pull request by FlorianKirmaier**
Original PR: https://github.com/javafxports/openjdk-jfx/pull/591

A side effect of JDK-8200224 is, that the first click on a JFXPanel is always a double click.
The cause for JDK-8200224 is the fix for JDK-8087914 (Clicking on Menu in JFXPanel ignored if another swing component has focus).
This fix introduced synthesized press-events, which is an unsafe fix for the problem.

The pull request introduces a new fix for the problem.
Instead of simulating new input-events, we set JavaFX Scene/Window to focused.
We do so, by using setFocused.
When the original Swing-Focus event is called slightly later, it won't have any side-effects,
because calling setFocused just sets the value of a boolean property.

I've now also added a unit-test for the problem.

## Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

## Issue
[JDK-8200224](https://bugs.openjdk.java.net/browse/JDK-8200224): First mouse press each time JFXPanel gains focus is triggered twice

## Approvers
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**) **Note!** Review applies to 24385eb8983d3600f36f64c4cbf24f1da3c84b8c
 * Prasanta Sadhukhan ([psadhukhan](@prsadhuk) - **Reviewer**) **Note!** Review applies to 24385eb8983d3600f36f64c4cbf24f1da3c84b8c

### CODE REVIEW

**Comment by kevinrushforth** [comment]
@FlorianKirmaier you still need to file a JBS issue to associate your GitHub username with your OpenJDK user ID as noted [above](#issuecomment-546883472), and per the instructions in [this message](https://mail.openjdk.java.net/pipermail/openjfx-dev/2019-September/023558.html) sent to openjfx-dev.

**Comment by kevinrushforth** [comment]
@prsadhuk please review this. I will also review it (this will need two reviewers).

File: modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java [hunk 1]
```diff
@@ -446,13 +446,11 @@ protected void processMouseEvent(MouseEvent e) {
...
                 requestFocus();
-                // this focus request event goes to eventqueue and will be
-                // asynchronously handled so MOUSE_PRESSED event will not be
-                // honoured by FX immediately due to lack of focus in fx
-                // component. Fire the same MOUSE_PRESSED event after
-                // requestFocus() so that 2nd mouse press will be honoured
-                // since now fx have focus
-                jfxPanelIOP.postEvent(this, e);
+                // This fixes JDK-8087914 without causing JDK-8200224
+                // It is save, because in JavaFX only the method "setFocused(true)" is called,
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java]: typo: save --> safe

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2014, 2016 Oracle and/or its affiliates. All rights reserved.
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: replace `2014, 2016` with `2019,`

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package test.javafx.embed.swing;
+
+import javafx.application.Application;
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: Remove unused import (several of the below imports are similarly unused). Also, since this is a new test, please sort the imports.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import javax.swing.JPanel;
+import javax.swing.JFrame;
+import javax.swing.SwingUtilities;
+
+import java.awt.Dimension;
+import java.awt.event.MouseEvent;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static junit.framework.Assert.assertEquals;
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This method should be imported from `org.junit` package.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                scene.addEventHandler(javafx.scene.input.MouseEvent.MOUSE_PRESSED, (event -> {
+                    pressedEventCounter[0] += 1;
+                    firstPressedEventLatch.countDown();
+                }));
+
+                fxPnl.setScene(scene);
+
+                SwingUtilities.invokeLater(() -> {
+                    MouseEvent e = new MouseEvent(fxPnl, MouseEvent.MOUSE_PRESSED, 0, MouseEvent.BUTTON1_DOWN_MASK,
+                            5, 5, 1, false, MouseEvent.BUTTON1);
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This doesn't appear to trigger the bug -- at least on Windows. The test passes for me with or without your fix. You might consider moving it to the system tests project, under `tests/system/src/test/java/test/robot` and using `Robot` to effect the mouse press.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import static junit.framework.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class JFXPanelTest {
+
+    // Used to launch the application before running any test
+    private static final CountDownLatch launchLatch = new CountDownLatch(1);
+
+
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: You can remove the extra blank line.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    @BeforeClass
+    public static void doSetupOnce() {
+        Platform.startup(() -> {
+            Platform.setImplicitExit(false);
+            launchLatch.countDown();
+        });
+
+
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: You can remove the extra blank line.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        } catch (InterruptedException ex) {
+            AssertionFailedError err = new AssertionFailedError("Unexpected exception");
+            err.initCause(ex);
+            throw err;
+        }
+
+        assertEquals(0, launchLatch.getCount());
+    }
+
+    class TestFXPanel extends JFXPanel {
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: I think this can be a static class.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        protected void processMouseEventPublic(MouseEvent e) {
+            processMouseEvent(e);
+        }
+    };
+
+    @Test
+    public void testNoDoubleClickOnFirstClick() throws Exception {
+
+        CountDownLatch firstPressedEventLatch = new CountDownLatch(1);
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This can be final.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+        // It's an array, so we can mutate it inside of lambda statement
+        int[] pressedEventCounter = {0};
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This can be final.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        if(!firstPressedEventLatch.await(5000, TimeUnit.MILLISECONDS)) {
+            throw new Exception();
+        };
+
+        Thread.sleep(100); // there should be no pressed event after the initial one. Let's wait for 0.1s and check again.
+
+        assertEquals(1, pressedEventCounter[0]);
+    }
+
+
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: You can remove the extra blank line.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+                SwingUtilities.invokeLater(() -> {
+                    MouseEvent e = new MouseEvent(fxPnl, MouseEvent.MOUSE_PRESSED, 0, MouseEvent.BUTTON1_DOWN_MASK,
+                            5, 5, 1, false, MouseEvent.BUTTON1);
+
+                    fxPnl.processMouseEventPublic(e);
+                });
+            });
+        });
+
+        if(!firstPressedEventLatch.await(5000, TimeUnit.MILLISECONDS)) {
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: Add a space after the `if`.

**Comment by FlorianKirmaier** [comment]
You can run the AWT_TESTS with the following statement:
```
./gradlew -PFULL_TEST=true swing:clean swing:test
```
For some reason, it's hidden behind this flag.
Maybe that's the reason, why you couldn't reproduce the bug?
As fas as I can thee, the swing-tests are stable, so there is no real reason to hide it behind a flag.
Maybe that's something which should be changed?

The tests are now stable. The previous version had some problems because the other test for Swing was calling System.exit.

The fix is well tested under Windows and Mac but not on Linux. The tests are based on a backport to an older JavaFX-version.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 5 above)
Comment (FlorianKirmaier) [reply modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: Maybe try `./gradlew -PFULL_TEST=true swing:clean swing:test`.
I'm sure, it can be reproduced on windows.
If you still can not reproduce it, then I will add a test for the Robot.

**Comment by kevinrushforth** [comment]
I'll try it again with the updated test.

> You can run the AWT_TESTS with the following statement:
> 
> ```
> ./gradlew -PFULL_TEST=true swing:clean swing:test
> ```

Yes, I know that it needs `-PFULL_TEST=true`, but the earlier test wasn't failing for me. And yes, it is intentional; for reasons I can't recall, the swing tests don't work in headless mode. Anyway, I don't want to revisit this right now.

> The tests are now stable. The previous version had some problems because the other test for Swing was calling System.exit.

Tests should not call `System.exit`. If they do, then that's almost certainly a bug.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/SwingFXUtilsTest.java [hunk 13]
```diff
@@ -43,46 +43,17 @@
 import org.junit.BeforeClass;
 import org.junit.Test;
 
+
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/SwingFXUtilsTest.java]: Your proposed fix for this test class is not the right fix, and needs to be reverted. It highlights an existing bug with the swing tests, which I will address in general comments.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        // It's an array, so we can mutate it inside of lambda statement
+        final int[] pressedEventCounter = {0};
+
+        SwingUtilities.invokeLater(() -> {
+            TestFXPanel fxPnl = new TestFXPanel();
+            fxPnl.setPreferredSize(new Dimension(100, 100));
+            JFrame jframe = new JFrame();
+            JPanel jpanel = new JPanel();
+            jpanel.add(fxPnl);
+            jframe.setContentPane(jpanel);
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: You should call `jframe.pack()` here.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+                    fxPnl.processMouseEventPublic(e);
+                });
+            });
+        });
+
+        if (!firstPressedEventLatch.await(5000, TimeUnit.MILLISECONDS)) {
+            throw new Exception("No first click detected.");
+        };
+
+        Thread.sleep(100); // there should be no pressed event after the initial one. Let's wait for 0.1s and check again.
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: I recommend using 500 msec so we don't risk missing a failing click on slower systems.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+public class JFXPanelTest {
+
+    // Used to launch the application before running any test
+    private static final CountDownLatch launchLatch = new CountDownLatch(1);
+    private static boolean setupIsDone = false;
+
+    @BeforeClass
+    public static void doSetupOnce() {
+        if (setupIsDone) return;
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This is not needed if you revert the changes to the other test class.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+            });
+        });
+
+        if (!firstPressedEventLatch.await(5000, TimeUnit.MILLISECONDS)) {
+            throw new Exception("No first click detected.");
+        };
+
+        Thread.sleep(100); // there should be no pressed event after the initial one. Let's wait for 0.1s and check again.
+        assertEquals(1, pressedEventCounter[0]);
+    }
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: You will want to add a cleanup method, annotated with `@AfterClass` to call `Platform.exit` and dispose of the JFrame.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+            JFrame jframe = new JFrame();
+            JPanel jpanel = new JPanel();
+            jpanel.add(fxPnl);
+            jframe.setContentPane(jpanel);
+            jframe.setVisible(true);
+
+            Platform.runLater(() -> {
+                Group grp = new Group();
+                Scene scene = new Scene(new Group());
+                scene.getRoot().requestFocus();
```
Comment (kevinrushforth) [inline modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: I don't think requesting focus is necessary (or desired).

**Comment by kevinrushforth** [comment]
FYI, I filed [JDK-8234110](https://bugs.openjdk.java.net/browse/JDK-8234110) to move the existing `:swing` test to `:systemTests`.

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/SwingFXUtilsTest.java (hunk 13 above)
Comment (FlorianKirmaier) [reply modules/javafx.swing/src/test/java/test/javafx/embed/swing/SwingFXUtilsTest.java]: It's now removed

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 15 above)
Comment (FlorianKirmaier) [reply modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: done

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 17 above)
Comment (FlorianKirmaier) [reply modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: I've kept the logic of the other JFXPanelTest. Should i change it?

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 16 above)
Comment (FlorianKirmaier) [reply modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: removed

File: modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 18 above)
Comment (FlorianKirmaier) [reply modules/javafx.swing/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: I've removed it

**Comment by FlorianKirmaier** [comment]
I've updated the pull request and merged it with master.
Later, I will retest it with a Windows VM.

**Comment by FlorianKirmaier** [comment]
I've done now the requested changes.
Now only retesting it with Windows is left.

**Comment by FlorianKirmaier** [comment]
I've now tested on Windows.
I can confirm, that the test doesn't reproduce the error on Windows.

**Comment by kevinrushforth** [comment]
Yes, I confirmed this last Friday, but ran out of time to reply. You also need to restore the test that you moved from `test.robot.*` to its original state (so we preserve the previous Mac-only test).

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: That should be `2019`

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+import com.sun.javafx.PlatformUtil;
+import org.junit.Assume;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.AfterClass;
+import org.junit.Test;
+import junit.framework.AssertionFailedError;
+
+
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: You only need one blank line here.

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+        Assert.assertEquals(0, launchLatch.getCount());
+    }
+
+    @AfterClass
+    public static void doTeardownOnce() {
+        Platform.exit();
+    }
+
+    class TestFXPanel extends JFXPanel {
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This class can be `static`

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Test
+    public void testNoDoubleClickOnFirstClick() throws Exception {
+
+        CountDownLatch firstPressedEventLatch = new CountDownLatch(1);
+
+        // It's an array, so we can mutate it inside of lambda statement
+        int[] pressedEventCounter = {0};
+
+        SwingUtilities.invokeLater(() -> {
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: Add the following here:
```
            JFXPanel dummyFXPanel = new JFXPanel();
            dummyFXPanel.setPreferredSize(new Dimension(100, 100));
```

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+            TestFXPanel fxPnl = new TestFXPanel();
+            fxPnl.setPreferredSize(new Dimension(100, 100));
+            JFrame jframe = new JFrame();
+            JPanel jpanel = new JPanel();
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: Add the following here:
```
            jpanel.add(dummyFXPanel);
```

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+            jpanel.add(fxPnl);
+            jframe.setContentPane(jpanel);
+            jframe.pack();
+            jframe.setVisible(true);
+
+            Platform.runLater(() -> {
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: Add the following here:
```
                Scene dummyScene = new Scene(new Group());
                dummyFXPanel.setScene(dummyScene);
```

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+                SwingUtilities.invokeLater(() -> {
+                    MouseEvent e = new MouseEvent(fxPnl, MouseEvent.MOUSE_PRESSED, 0, MouseEvent.BUTTON1_DOWN_MASK,
+                            5, 5, 1, false, MouseEvent.BUTTON1);
+
+                    fxPnl.processMouseEventPublic(e);
+                });
+            });
+        });
+
+        if(!firstPressedEventLatch.await(5000, TimeUnit.MILLISECONDS)) {
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This `if` test can be written more succinctly as:

```
    assertTrue(firstPressedEventLatch.await(5000, TimeUnit.MILLISECONDS));
```

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+            Platform.setImplicitExit(false);
+            Assert.assertTrue(Platform.isFxApplicationThread());
+            Assert.assertNotNull(primaryStage);
+
+            launchLatch.countDown();
+        }
+    }
+
+    @BeforeClass
+    public static void doSetupOnce() {
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: If you add `throws Exception` then you don't need the try/catch around the `launchLatch.await`.

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+            fxPnl.setPreferredSize(new Dimension(100, 100));
+            JFrame jframe = new JFrame();
+            JPanel jpanel = new JPanel();
+            jpanel.add(fxPnl);
+            jframe.setContentPane(jpanel);
+            jframe.pack();
+            jframe.setVisible(true);
+
+            Platform.runLater(() -> {
+                Group grp = new Group();
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: This variable is unused.

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 25 above)
Comment (FlorianKirmaier) [reply tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: The if is triggered, when the time runs out.

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 27 above)
Comment (FlorianKirmaier) [reply tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: remvoed

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 20 above)
Comment (FlorianKirmaier) [reply tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: done

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 22 above)
Comment (FlorianKirmaier) [reply tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: done

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 26 above)
Comment (FlorianKirmaier) [reply tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: It's now simplified.

File: tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java (hunk 24 above)
Comment (FlorianKirmaier) [reply tests/system/src/test/java/test/javafx/embed/swing/JFXPanelTest.java]: done

**Comment by FlorianKirmaier** [comment]
1. I've restored the test. But the git history is now very chaotic. Especially the moves might cause problems. Do the commits get squashed after merging? Otherwise, it might make sense, to redo the changes on a fresh branch and create a new pull request.
2. The test works now on windows. : )

**Comment by kevinrushforth** [comment]
Yes, the commits are squashed, so don't worry about the history (in worst case I would ask you to do a squash-rebase / force push rather than create a new PR, but that isn't needed here).

At first glance the changes you made look good. I'll take a closer look later.

@prsadhuk please also review this.

File: modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java [hunk 28]
```diff
@@ -446,13 +446,11 @@ protected void processMouseEvent(MouseEvent e) {
...
             if (isFocusable() && !hasFocus()) {
                 requestFocus();
-                // this focus request event goes to eventqueue and will be
-                // asynchronously handled so MOUSE_PRESSED event will not be
-                // honoured by FX immediately due to lack of focus in fx
-                // component. Fire the same MOUSE_PRESSED event after
-                // requestFocus() so that 2nd mouse press will be honoured
-                // since now fx have focus
-                jfxPanelIOP.postEvent(this, e);
+                // This fixes JDK-8087914 without causing JDK-8200224
```
Comment (prsadhuk) [inline modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java]: We actually do not clutter source code with bugids, it will be good if it can be removed with proper comments maybe something like "extra simulated mouse pressed event is removed by making the JavaFX scene focused"

Comment (kevinrushforth) [reply modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java]: In general, we no longer reference bug IDs in source code, so I agree with the recommendation to reword the comment.

@FlorianKirmaier - can you reword as suggested? Once you are done, you can integrate it (using the `/integrate` command), and I will sponsor it. There will likely be a delay due to the US Thanksgiving holiday.

Comment (FlorianKirmaier) [reply modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java]: Done
